### PR TITLE
Refactor: improve code quality, security, and add occurrence validation

### PR DESCRIPTION
**Pull request by andrest50**
## Summary

This PR includes several refactoring improvements, security enhancements, and new features:

- **Security Enhancement**: Fixed Zoom webhook signature validation bypass vulnerability by making signature headers required
- **Architecture Improvement**: Refactored webhook logic from API layer to service layer following clean architecture principles
- **Privacy Protection**: Implemented comprehensive PII redaction for email addresses in logging throughout the codebase
- **Performance Optimization**: Replaced mutex-protected counters with atomic operations and optimized meeting data fetching
- **Code Quality**: Reduced cyclomatic complexity of doRequest function from 25 to 8 through meaningful refactoring
- **Feature Addition**: Added occurrence ID validation for meeting registrants to ensure valid and future-only registrations
- **Bug Fix**: Fixed weekly recurrence calculation that was generating past occurrences when meeting start day differed from recurrence day

## Jira Tickets

- [LFXV2-390](https://linuxfoundation.atlassian.net/browse/LFXV2-390)
- [LFXV2-336](https://linuxfoundation.atlassian.net/browse/LFXV2-336)  
- [LFXV2-446](https://linuxfoundation.atlassian.net/browse/LFXV2-446)

## Test plan

- [x] All existing tests pass
- [x] Added comprehensive test coverage for occurrence validation
- [x] Added test case for weekly recurrence edge case
- [x] Verified race conditions are resolved with race detector
- [x] Confirmed application builds and compiles successfully
- [x] Tested webhook signature validation in both production and mock modes

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: internal/infrastructure/zoom/api/client.go [hunk 1]
```diff
@@ -189,170 +189,224 @@ func (c *Client) calculateBackoff(attempt int) time.Duration {
...
+			c.logSuccessfulResponse(ctx, method, path, resp, duration, attempt)
+			return resp, nil
 		}
 
-		// Check if we should retry
-		statusCode := 0
-		if resp != nil {
-			statusCode = resp.StatusCode
-		}
+		// Update state and determine if we should retry
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/infrastructure/zoom/api/client.go:218]: Fixed.

File: internal/handlers/zoom_webhook_handlers.go [hunk 2]
```diff
@@ -1209,22 +1211,21 @@ func (s *ZoomWebhookHandler) createPastMeetingParticipants(ctx context.Context,
...
 					"past_meeting_uid", pastMeeting.UID,
 					"meeting_uid", meeting.UID,
-					"email", r.Email,
+					"email", redaction.RedactEmail(r.Email),
 				)
-				failedCount++
+				atomic.AddInt64(&failedCount, 1)
+				// Use mutex only for slice access
+				mu.Lock()
 				failedEmails = append(failedEmails, r.Email)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/handlers/zoom_webhook_handlers.go:1224]: Fixed.

File: cmd/meeting-api/api.go [hunk 3]
```diff
@@ -154,132 +152,51 @@ func (s *MeetingsAPI) JWTAuth(ctx context.Context, bearerToken string, _ *securi
...
-	if !s.zoomWebhookHandler.HandlerReady() {
-		logger.ErrorContext(ctx, "Service not ready")
-		return nil, createResponse(http.StatusServiceUnavailable, domain.NewUnavailableError("service unavailable", nil))
+	// Log webhook info with redacted payload to protect PII
+	slog.InfoContext(ctx, "Zoom webhook received",
+		"event_type", payload.Event,
+		"event_ts", payload.EventTs,
+		"payload", payload.Payload,
+	)
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/meeting-api/api.go:165]: It's fine as is for now.

File: cmd/meeting-api/platforms/zoom.go [hunk 4]
```diff
@@ -66,12 +68,16 @@ func SetupZoom(config ZoomConfig) ZoomConfig {
...
+		slog.Warn("MOCK_ZOOM_WEBHOOK is enabled - webhook signature validation will be bypassed")
+		config.Validator = webhook.NewMockWebhookValidator()
+	case config.WebhookSecretToken != "":
 		validator := webhook.NewZoomWebhookValidator(config.WebhookSecretToken)
 		slog.Info("Zoom webhook validation configured")
 		config.Validator = validator
-	} else {
+	default:
 		slog.Warn("Zoom webhook validation not configured", logging.PriorityCritical())
 	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/meeting-api/platforms/zoom.go:82]: When mock is enabled then this is fine.

File: cmd/meeting-api/main.go [hunk 5]
```diff
@@ -121,6 +122,10 @@ func main() {
 		registrantService,
 		messageBuilder,
 	)
+	zoomWebhookService := service.NewZoomWebhookService(
+		messageBuilder,
+		platformConfigs.Zoom.Validator,
+	)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/meeting-api/main.go:129]: It's fine as it is.

File: internal/infrastructure/zoom/api/client.go [hunk 6]
```diff
@@ -189,170 +189,224 @@ func (c *Client) calculateBackoff(attempt int) time.Duration {
...
-				resp.Body = io.NopCloser(bytes.NewReader(body))
-				slog.ErrorContext(ctx, "Zoom API error response",
-					"method", method,
-					"path", path,
-					"status", resp.StatusCode,
-					"duration", duration.String(),
-					"body", string(body),
-					logging.ErrKey, fmt.Errorf("status: %d", resp.StatusCode))
-			}
+		c.logRequestAttempt(ctx, method, path, body, attempt)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/infrastructure/zoom/api/client.go:208]: It's fine for now.

File: internal/service/zoom_webhook_service.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+	// Generate encrypted token using HMAC SHA-256
+	// Cast to ZoomWebhookValidator to get the secret token
+	zoomValidator, ok := s.WebhookValidator.(*webhook.ZoomWebhookValidator)
+	if !ok || zoomValidator.SecretToken == "" {
+		logger.ErrorContext(ctx, "Zoom webhook validator not properly configured")
+		return nil, domain.NewInternalError("webhook validation not configured")
+	}
+
+	h := hmac.New(sha256.New, []byte(zoomValidator.SecretToken))
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/zoom_webhook_service.go:131]: Fixed.

File: internal/infrastructure/zoom/api/client.go [hunk 8]
```diff
@@ -189,170 +189,224 @@
...
+		_ = resp.Body.Close()
+		resp.Body = io.NopCloser(bytes.NewReader(body))
+		slog.ErrorContext(ctx, "Zoom API error response",
+			"method", method,
+			"path", path,
+			"status", resp.StatusCode,
+			"duration", duration.String(),
+			"body", string(body),
+			logging.ErrKey, fmt.Errorf("status: %d", resp.StatusCode))
+	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/infrastructure/zoom/api/client.go:338]: This is okay for now.

File: internal/infrastructure/zoom/api/client.go [hunk 9]
```diff
@@ -189,170 +189,224 @@
...
+}
+
+// closeAndReplaceResponse closes the old response if it exists and returns the new one
+func (c *Client) closeAndReplaceResponse(oldResp, newResp *http.Response) *http.Response {
+	if oldResp != nil && newResp != nil {
+		_ = oldResp.Body.Close()
+	}
+	return newResp
+}
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/infrastructure/zoom/api/client.go:307]: Fixed.

File: internal/infrastructure/zoom/api/client.go [hunk 10]
```diff
@@ -189,170 +189,224 @@
...
-			} else {
-				slog.ErrorContext(ctx, "Zoom API request failed (not retryable)",
-					"method", method,
-					"path", path,
-					"status", statusCode,
-					"duration", duration.String(),
-					"attempt", attempt+1)
-			}
+			c.logNonRetryableError(ctx, method, path, statusCode, duration, attempt, err)
 			break
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/infrastructure/zoom/api/client.go:224]: Fixed.

File: internal/handlers/zoom_webhook_handlers.go [hunk 11]
```diff
@@ -492,7 +494,7 @@ func (s *ZoomWebhookHandler) handleMeetingEndedEvent(ctx context.Context, event
 
 	// Try to find existing PastMeeting record by platform meeting ID and occurrence ID
 	existingPastMeeting, err := s.pastMeetingService.PastMeetingRepository.GetByPlatformMeetingIDAndOccurrence(ctx, models.PlatformZoom, meetingObj.ID, occurrenceID)
-	if err != nil && err != domain.NewNotFoundError("past meeting not found", nil) {
+	if err != nil && err != domain.NewNotFoundError("past meeting not found") {
 		slog.ErrorContext(ctx, "error searching for existing past meeting", logging.ErrKey, err)
 		return fmt.Errorf("failed to search for existing past meeting: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/handlers/zoom_webhook_handlers.go:500]: Fixed.

File: cmd/meeting-api/api_meeting_registrants.go [hunk 12]
```diff
@@ -14,7 +14,7 @@ import (
 // GetMeetingRegistrants gets all meeting registrants for a meeting.
 func (s *MeetingsAPI) GetMeetingRegistrants(ctx context.Context, payload *meetingsvc.GetMeetingRegistrantsPayload) (*meetingsvc.GetMeetingRegistrantsResult, error) {
 	if payload == nil || payload.UID == nil {
-		return nil, domain.NewValidationError("validation failed", nil)
+		return nil, domain.NewValidationError("validation failed")
 	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/meeting-api/api_meeting_registrants.go:18]: Fixed.

**Review by prabodhcs — DISMISSED** [review]
Good refactoring.
